pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -1008,7 +1030,8 @@ public void stop()
   // If the transition is OFFLINE to ONLINE, the caller should have downloaded the segment and we don't reach here.
   public LLRealtimeSegmentDataManager(RealtimeSegmentZKMetadata segmentZKMetadata, TableConfig tableConfig,
       InstanceZKMetadata instanceZKMetadata, RealtimeTableDataManager realtimeTableDataManager, String resourceDataDir,
-      IndexLoadingConfig indexLoadingConfig, Schema schema, ServerMetrics serverMetrics) {
+      IndexLoadingConfig indexLoadingConfig, Schema schema, Map<Integer, Semaphore> partitionIdToSemaphoreMap,
```

Comment:
you do not need to pass the entire map. Just one semaphore should be enough. The caller can call getOrDefault()

Reply:
The problem is partitionId is extracted inside LLRealtimeSegmentDataManager.

Reply:
you could introduce `getSemaphoreForPartition` method in the RealtimeDataManager, similar to `getSegmentBuildSemaphore`.  Inside getSemaphoreForPartition, concurrentMap::compute can be used to either return existing value or new semaphore if null. That way you don't have to pass the whole map.

Reply:
Good idea. But the creation of Kafka consumers happens in the constructor of RealtimeDataManager, which means the consumers have already been created before the SegmentManager acquires the semaphore.

Reply:
You can get the partiitionId in the calling method from the SegmentName object. Passing the entire map into this object is not clean

Reply:
Updated the code to pass the semaphore only to segmentManager.